### ISSUE TITLE
Can we remove -rdynamic?

### ISSUE THREAD

**Issue by nuko8** [awesomeWM/awesome#450]
Currently, the compiler option `-rdynamic` is set at `awesomeConfig.cmake:19` with `add_definitions()`.

As a result, the option is passed to the compiler every time it compiles a source file to yield the corresponding object file, but never when it links the resulting object files.

Actually, observing the output of `make VERBOSE=1`, we see that the option is not passed to the compiler at the link phase.

On the other hand, according to the GCC manual, [3.13 Options for Linking](https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html), `-rdynamic` is one of such options that

>These options come into play when the compiler links object files into an executable output file. They are meaningless if the compiler is not doing a link step.

To sum up, `-rdynamic` fails to be relevant to the build.  Therefore, it should be safe to remove it.

Or, is the option indeed working to add all the symbols to the dynamic symbol table to help backtrace?  I cannot check if this is true or not for myself because I'm using a compiler which does not support the option.

I'd like to confirm whether or not `-rdynamic` is inevitable for systems which support it.

**Comment by psychon** [comment]
Make awesome crash at startup:
```diff
diff --git a/root.c b/root.c
index 59534fc..79aec8b 100644
--- a/root.c
+++ b/root.c
@@ -37,12 +37,14 @@
 #include <xcb/xcb_aux.h>
 #include <cairo-xcb.h>
 
+#include <signal.h>
 static void
 root_set_wallpaper_pixmap(xcb_connection_t *c, xcb_pixmap_t p)
 {
     xcb_get_property_cookie_t prop_c;
     xcb_get_property_reply_t *prop_r;
     const xcb_screen_t *screen = globalconf.screen;
+    raise(SIGSEGV);
 
     /* We now have the pattern painted to the pixmap p. Now turn p into the root
      * window's background pixmap.
```

Output:
```
E: awesome: signal_fatal:308: signal 11, dumping backtrace
./awesome(backtrace_get+0x41) [0x422834]
./awesome() [0x40dd5f]
/lib/x86_64-linux-gnu/libc.so.6(+0x35180) [0x7fecd26a1180]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x37) [0x7fecd26a1107]
./awesome() [0x41ca80]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xcc4d) [0x7fecd2f72c4d]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x1883d) [0x7fecd2f7e83d]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xcf8e) [0x7fecd2f72f8e]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xc5cf) [0x7fecd2f725cf]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xd1d1) [0x7fecd2f731d1]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(lua_pcallk+0x71) [0x7fecd2f6f0f1]
./awesome() [0x4160d3]
./awesome(luaA_parserc+0x25) [0x41996d]
./awesome(main+0x8f2) [0x40e863]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7fecd268db45]
./awesome() [0x40dac9]
```

After removing `-rdynamic` and a (hopefully correct) rebuild, I get pretty much the same output.

However, Google suggests that CMake hardcodes `-rdynamic` where it can. After adding

```
SET(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS)
SET(CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS)
```
(and removing random files in `build/` until CMake stopped passing `-rdynamic`) I now get the following backtrace:

```
E: awesome: signal_fatal:308: signal 11, dumping backtrace
./awesome() [0x41de94]
./awesome() [0x4093bf]
/lib/x86_64-linux-gnu/libc.so.6(+0x35180) [0x7fa9f0e29180]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x37) [0x7fa9f0e29107]
./awesome() [0x4180e0]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xcc4d) [0x7fa9f16fac4d]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0x1883d) [0x7fa9f170683d]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xcf8e) [0x7fa9f16faf8e]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xc5cf) [0x7fa9f16fa5cf]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(+0xd1d1) [0x7fa9f16fb1d1]
/usr/lib/x86_64-linux-gnu/liblua5.2.so.0(lua_pcallk+0x71) [0x7fa9f16f70f1]
./awesome() [0x411733]
./awesome() [0x414fcd]
./awesome() [0x409ec3]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7fa9f0e15b45]
./awesome() [0x409129]
```

**Comment by nuko8** [comment]
Thank you for taking your time to investigate it.  That gave me a clue to investigate it further.

In the CMake module file `/.../share/cmake-x.y/Platform/Linux-GNU.cmake`, CMake defines a macro in which `CMAKE_SHARED_LIBRARY_LINK_${lang}_FLAGS` is set to `-rdynamic`.  CMake implicitly uses the macro at the link phase.

That's why `set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS "")` makes difference in traceback while removing `-rdynamic` from `awesomeConfig.cmake` not much (on Linux with GCC).

We may conclude that the current `awesomeConfig.cmake` fails to pass `-rdynamic` to the linker correctly, and that CMake implicitly does the job instead.

So, how about removing `-rdynamic` from the parameter of `add_definitions()` and add it to the linker flag via, e.g., `target_link_libraries()` or `set_target_properties(...PROPERTIES LINK_FLAGS ...)` for Linux with GCC?

**Comment by psychon** [comment]
Since CMake adds it anyway, we may as well just remove it. If that breaks, we can then add it properly. I also don't mind doing it properly right now either. :-)

**Comment by nuko8** [comment]
That's just the answer that I think is the best!

> I also don't mind doing it properly right now either. :-)

I do appreciate it, of course.

**Comment by nuko8** [comment]
Closed as the relevant PR #453 was closed.
